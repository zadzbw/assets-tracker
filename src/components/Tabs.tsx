import { type FC, type ReactNode, useRef, useState, useEffect, Fragment } from 'react'
import { clsx } from 'clsx'

interface TabItemType {
  tabKey: string
  label: ReactNode
  children: ReactNode
  destroyOnHidden?: boolean // 被隐藏时是否销毁 DOM 结构，优先级比 forceRender 更高
  forceRender?: boolean // 被隐藏时是否渲染 DOM 结构
}

interface TabNavType extends Pick<TabItemType, 'tabKey' | 'label'> {
  isActive: boolean
  onChange: (activeKey: string) => void
}

interface TabContentType extends Omit<TabItemType, 'tabKey' | 'label'> {
  isActive: boolean
}

interface BaseTabsProps {
  items: TabItemType[]
}

interface ControlledTabsProps extends BaseTabsProps {
  activeKey: string
  onChange: (activeKey: string) => void
}

interface UnControlledTabsProps extends BaseTabsProps {
  defaultActiveKey: string
}

type TabsProps = ControlledTabsProps | UnControlledTabsProps

const TabNav: FC<TabNavType> = ({ isActive, label, tabKey, onChange }) => {
  return (
    <div
      className={clsx('tab', { 'tab-active': isActive })}
      onClick={() => {
        if (!isActive) {
          onChange(tabKey)
        }
      }}
    >
      {label}
    </div>
  )
}

const TabContent: FC<TabContentType> = ({
  isActive,
  children,
  destroyOnHidden = false,
  forceRender = false,
}) => {
  const isMounted = useRef<boolean>(false)

  const content = (
    <div
      ref={() => {
        isMounted.current = true
      }}
      className="tab-content border-base-300 bg-base-100 p-4"
    >
      {children}
    </div>
  )

  if (destroyOnHidden) {
    return isActive && content
  }
  if (forceRender) {
    return content
  }
  if (isMounted.current) {
    return content
  }
  return isActive && content
}

export const Tabs: FC<TabsProps> = (props) => {
  const { items } = props
  const isControlled = 'activeKey' in props

  // 组件内部 state
  const [activeKey, setActiveKey] = useState(() =>
    isControlled ? props.activeKey : props.defaultActiveKey,
  )

  const handleChange = (key: string) => {
    if (isControlled) {
      props.onChange(key)
    } else {
      setActiveKey(key)
    }
  }

  // 受控组件，props.activeKey 如果变更，需要更新内部 state
  useEffect(() => {
    if (isControlled) {
      setActiveKey(props.activeKey)
    }
  }, [isControlled, isControlled && props.activeKey])

  return (
    <div className="tabs tabs-border">
      {items.map(({ tabKey, label, ...rest }) => {
        const isActive = tabKey === activeKey
        return (
          <Fragment key={tabKey}>
            <TabNav isActive={isActive} tabKey={tabKey} label={label} onChange={handleChange} />
            <TabContent isActive={isActive} {...rest} />
          </Fragment>
        )
      })}
    </div>
  )
}
