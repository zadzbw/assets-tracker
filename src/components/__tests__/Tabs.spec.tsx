import { useState } from 'react'
import { describe, it, expect } from 'vitest'
import { render } from 'vitest-browser-react'
import { Tabs } from '../Tabs'

const items = [
  {
    tabKey: '1',
    label: 'Tab 1',
    children: 'Tab Content 1',
  },
  {
    tabKey: '2',
    label: 'Tab 2',
    children: 'Tab Content 2',
  },
  {
    tabKey: '3',
    label: 'Tab 3',
    children: 'Tab Content 3',
  },
]

describe('`Tabs` component test', () => {
  it('should render active tab content', async () => {
    const { container, getByText } = render(<Tabs items={items} defaultActiveKey="2" />)

    const tabContent1 = getByText('Tab Content 1')
    const tabContent2 = getByText('Tab Content 2')
    const tabContent3 = getByText('Tab Content 3')

    const tab2 = getByText('Tab 2')

    await expect.element(tabContent1).not.toBeInTheDocument()
    await expect.element(tabContent2).toBeInTheDocument()
    await expect.element(tabContent3).not.toBeInTheDocument()

    await expect.element(tabContent2).toHaveClass('tab-content')
    await expect.element(tab2).toHaveClass('tab tab-active')

    expect(container.firstElementChild).toMatchSnapshot()
  })

  it('should switch tab content', async () => {
    const { getByText } = render(<Tabs items={items} defaultActiveKey="2" />)

    await getByText('Tab 1').click()
    await expect.element(getByText('Tab Content 1')).toBeInTheDocument()

    await getByText('Tab 3').click()
    await expect.element(getByText('Tab Content 3')).toBeInTheDocument()
  })

  it('should keep DOM of hidden tab(opened previously)', async () => {
    const { getByText } = render(<Tabs items={items} defaultActiveKey="2" />)

    // 最开始没有 Tab Content 1 的 DOM
    await expect.element(getByText('Tab Content 1')).not.toBeInTheDocument()

    await getByText('Tab 1').click()
    await expect.element(getByText('Tab 1')).toHaveClass('tab tab-active')
    await expect.element(getByText('Tab Content 1')).toBeInTheDocument()

    await getByText('Tab 2').click()
    await expect.element(getByText('Tab 2')).toHaveClass('tab tab-active')
    // 即使切换到 2 以后，Tab Content 1 的 DOM 也依然存在
    await expect.element(getByText('Tab Content 1')).toBeInTheDocument()
  })

  it('should destroy DOM of hidden tab', async () => {
    const { getByText } = render(
      <Tabs
        items={[
          {
            tabKey: '1',
            label: 'Tab 1',
            children: 'Tab Content 1',
            destroyOnHidden: true,
          },
          {
            tabKey: '2',
            label: 'Tab 2',
            children: 'Tab Content 2',
          },
        ]}
        defaultActiveKey="2"
      />,
    )

    // 最开始没有 Tab Content 1 的 DOM
    await expect.element(getByText('Tab Content 1')).not.toBeInTheDocument()

    await getByText('Tab 1').click()
    await expect.element(getByText('Tab 1')).toHaveClass('tab tab-active')
    await expect.element(getByText('Tab Content 1')).toBeInTheDocument()
    await expect.element(getByText('Tab Content 2')).toBeInTheDocument()

    await getByText('Tab 2').click()
    await expect.element(getByText('Tab 2')).toHaveClass('tab tab-active')
    // 切换到 2 以后，Tab Content 1 的 DOM 被销毁
    await expect.element(getByText('Tab Content 1')).not.toBeInTheDocument()
  })

  it('should force render DOM of hidden tab', async () => {
    const { getByText } = render(
      <Tabs
        items={[
          {
            tabKey: '1',
            label: 'Tab 1',
            children: 'Tab Content 1',
            forceRender: true,
          },
          {
            tabKey: '2',
            label: 'Tab 2',
            children: 'Tab Content 2',
          },
        ]}
        defaultActiveKey="2"
      />,
    )

    // 最开始 Tab Content 1 的 DOM 也会存在
    await expect.element(getByText('Tab Content 1')).toBeInTheDocument()
    await expect.element(getByText('Tab Content 2')).toBeInTheDocument()
  })

  it('Controlled - should switch tab content', async () => {
    const Test = () => {
      const [key, setKey] = useState('1')
      return <Tabs items={items} activeKey={key} onChange={setKey} />
    }

    const { getByText } = render(<Test />)

    await getByText('Tab 1').click()
    await expect.element(getByText('Tab Content 1')).toBeInTheDocument()

    await getByText('Tab 3').click()
    await expect.element(getByText('Tab Content 3')).toBeInTheDocument()
  })
})
