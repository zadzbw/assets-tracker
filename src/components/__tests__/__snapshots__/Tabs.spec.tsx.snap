// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`\`Tabs\` component test > should render active tab content 1`] = `
<div
  class="tabs tabs-border"
>
  <div
    class="tab"
  >
    Tab 1
  </div>
  <div
    class="tab tab-active"
  >
    Tab 2
  </div>
  <div
    class="tab-content border-base-300 bg-base-100 p-4"
  >
    Tab Content 2
  </div>
  <div
    class="tab"
  >
    Tab 3
  </div>
</div>
`;
